### PR TITLE
yosys: add support for loading plugins, revert symbiflow changes

### PR DESCRIPTION
**Pull request by thoughtpolice**
See the discussion in [#105709](https://github.com/NixOS/nixpkgs/pull/105709#issuecomment-760692114).

This includes new `yosys-ghdl` and `yosys-bluespec` plugins as examples of how to use this infrastructure. Includes a bump to `ghdl` as well, and some minor tweaks to yosys.

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Please split ``init: experimental yosys-{bluespec, ghdl} plugins`` into two commits: ``yosys-bluespec: init at X``.

File: pkgs/development/compilers/ghdl/default.nix [hunk 1]
```diff
@@ -5,13 +5,15 @@ assert backend == "mcode" || backend == "llvm";
...
+  version = "unstable-2021.01.14";
 
   src = fetchFromGitHub {
-    owner = "ghdl";
-    repo = "ghdl";
-    rev = "v${version}";
-    sha256 = "0b53yl4im33c1cd4mdyc4ks9cmrpixym17gzchfmplrl22w3l17y";
+    owner  = "ghdl";
+    repo   = "ghdl";
+    rev    = "4868294436574660552ccef50a5b0849559393de";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/ghdl/default.nix:15]:
```suggestion
    owner = "ghdl";
    repo = "ghdl";
    rev = "4868294436574660552ccef50a5b0849559393de";
```

File: pkgs/development/compilers/ghdl/default.nix [hunk 2]
```diff
@@ -23,7 +25,7 @@ stdenv.mkDerivation rec {
     sed -i 's/check_version 7.0/check_version 7/g' configure
   '';
 
-  configureFlags = lib.optional (backend == "llvm")
+  configureFlags = [ "--enable-synth" ] ++ lib.optional (backend == "llvm")
     "--with-llvm-config=${llvm}/bin/llvm-config";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/ghdl/default.nix:29]:
```suggestion
  configureFlags = [
    "--enable-synth"
  ] ++ lib.optional (backend == "llvm") "--with-llvm-config=${llvm}/bin/llvm-config";
```

File: pkgs/development/compilers/yosys/default.nix [hunk 3]
```diff
@@ -46,17 +43,18 @@ stdenv.mkDerivation rec {
   };
 
   enableParallelBuilding = true;
-  nativeBuildInputs = [ pkg-config protobuf flex bison python3 ];
-  buildInputs = [ tcl readline libffi python3 protobuf zlib ];
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [ tcl readline libffi python3 bison flex protobuf zlib ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/default.nix:47]:
```suggestion
  nativeBuildInputs = [ bison flex pkg-config ];
  buildInputs = [ tcl readline libffi python3 protobuf zlib ];
```

File: pkgs/development/compilers/yosys/default.nix [hunk 4]
```diff
@@ -46,17 +43,18 @@ stdenv.mkDerivation rec {
...
 
   makeFlags = [ "ENABLE_PROTOBUF=1" "PREFIX=${placeholder "out"}"];
 
   patchPhase = ''
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/default.nix:51]:
```suggestion
  postPatch = ''
```

File: pkgs/development/compilers/yosys/default.nix [hunk 5]
```diff
@@ -92,14 +90,9 @@ stdenv.mkDerivation rec {
   # add a symlink to fake things so that both variants work the same way. this
   # is also needed at build time for the test suite.
   postBuild   = "ln -sfv ${abc-verifier}/bin/abc ./yosys-abc";
-  postInstall = ''
-    ln -sfv ${abc-verifier}/bin/abc $out/bin/yosys-abc
+  postInstall = "ln -sfv ${abc-verifier}/bin/abc $out/bin/yosys-abc";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/default.nix:93]:
```suggestion
  postInstall = ''
    ln -sfv ${abc-verifier}/bin/abc $out/bin/yosys-abc
  '';
```

File: pkgs/development/compilers/yosys/plugins/bluespec.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+}:
+
+stdenv.mkDerivation {
+  pname = "yosys-bluespec";
+  version = "2021.01.14";
+
+  src = fetchFromGitHub {
+    owner  = "thoughtpolice";
+    repo   = "yosys-bluespec";
+    rev    = "7a0c609e9b601560067c34b8ce41e89ea4a397d4";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/plugins/bluespec.nix:12]:
```suggestion
    owner = "thoughtpolice";
    repo = "yosys-bluespec";
    rev = "7a0c609e9b601560067c34b8ce41e89ea4a397d4";
```

File: pkgs/development/compilers/yosys/plugins/bluespec.nix [hunk 7]
```diff
@@ -0,0 +1,32 @@
+{ stdenv, fetchFromGitHub, pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/plugins/bluespec.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, pkg-config
```

File: pkgs/development/compilers/yosys/plugins/bluespec.nix [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  src = fetchFromGitHub {
+    owner  = "thoughtpolice";
+    repo   = "yosys-bluespec";
+    rev    = "7a0c609e9b601560067c34b8ce41e89ea4a397d4";
+    sha256 = "0xj9j5dwvq0dwqzfjd4ikmxkd28amj6d9m6ava7pwcrpjhyjnd0c";
+  };
+
+  buildInputs = [ yosys readline zlib bluespec ];
+  nativeBuildInputs = [ pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/plugins/bluespec.nix:17]:
```suggestion
  nativeBuildInputs = [ pkg-config ];
  buildInputs = [ yosys readline zlib bluespec ];
```

File: pkgs/development/compilers/yosys/plugins/bluespec.nix [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  makeFlags = [
+    "PREFIX=$(out)/share/yosys/plugins"
+    "STATIC_BSC_PATH=${bluespec}/bin/bsc"
+    "STATIC_BSC_LIBDIR=${bluespec}/lib"
+  ];
+
+  meta = with stdenv.lib; {
+    description = "Bluespec plugin for Yosys";
+    license     = licenses.isc;
+    platforms   = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/plugins/bluespec.nix:29]: I am not a fan of this indentation.

File: pkgs/development/compilers/yosys/plugins/ghdl.nix [hunk 10]
```diff
@@ -0,0 +1,31 @@
+{ stdenv, fetchFromGitHub, pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/plugins/ghdl.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, pkg-config
```

File: pkgs/development/compilers/yosys/plugins/ghdl.nix [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+  src = fetchFromGitHub {
+    owner  = "ghdl";
+    repo   = "ghdl-yosys-plugin";
+    rev    = "8b3e740fef9a653a20a838f1ade475960e1d379b";
+    sha256 = "13kpcp37r11xgqd8aksddi40a5vpzqi7p8qgyjbs99lyscyli75x";
+  };
+
+  buildInputs = [ yosys readline zlib ghdl ];
+  nativeBuildInputs = [ pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/yosys/plugins/ghdl.nix:17]:
```suggestion
  nativeBuildInputs = [ pkg-config ];
  buildInputs = [ yosys readline zlib ghdl ];
```

File: pkgs/top-level/python-packages.nix [hunk 12]
```diff
@@ -7405,8 +7403,6 @@ in {
 
   sybil = callPackage ../development/python-modules/sybil { };
 
-  symbiflow-fasm = callPackage ../development/python-modules/symbiflow-fasm { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:7408]: Can we leave those packages in?

File: pkgs/development/compilers/yosys/default.nix [hunk 13]
```diff
@@ -43,23 +43,18 @@ stdenv.mkDerivation rec {
...
 
   makeFlags = [ "ENABLE_PROTOBUF=1" "PREFIX=${placeholder "out"}"];
 
   patchPhase = ''
     patch -p1 < ${./plugin-search-dirs.patch}
     substituteInPlace ./Makefile \
-      --replace 'CXX = clang' "" \
-      --replace 'LD = clang++' 'LD = $(CXX)' \
-      --replace 'CXX = gcc' "" \
-      --replace 'LD = gcc' 'LD = $(CXX)' \
```
Comment (CajuM) [inline pkgs/development/compilers/yosys/default.nix]: As I recall it's necessary to use clang++ or g++ as LD.

**Comment by thoughtpolice** [comment]
> Please split init: experimental yosys-{bluespec, ghdl} plugins into two commits like yosys-bluespec: init at X.

Sorry, but I don't see a point in this at all. In fact I would argue my current version is superior: the singular commit gives complete rational for the change and its design (vs scoping plugins in a different attrset or lambda) and splitting the two is pointless in light of this, even if it's a more standard form. That said, the commit message could still be better.

> TLDR: Why did you revert the merge? Maybe just revert the individual commits?

Sorry, but no. I'm not about to get into debates about the merits of revert styles (of all things...), but that aside, I'll just ask: why would I want to revert 15 individual commits instead of just reverting 1?

File: pkgs/development/compilers/ghdl/default.nix (hunk 1 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/ghdl/default.nix:15]: No.

File: pkgs/development/compilers/yosys/plugins/bluespec.nix (hunk 9 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/plugins/bluespec.nix:29]: That is, and I say this respectfully, not my problem.

File: pkgs/top-level/python-packages.nix (hunk 12 above)
Comment (CajuM) [reply pkgs/top-level/python-packages.nix:7408]: I think only textx is usefull without Symbiflow and project XRay

Comment (thoughtpolice) [reply pkgs/top-level/python-packages.nix:7408]: No. Please read my actual motivation beforehand, since you seemed to have missed it (I linked it in the PR description). Individual packages like `symbiflow-fasm` can go elsewhere (e.g. in my proposed separate repository, or NUR), and if  CajuM/Mihai wants to re-add some of the low-maintenance packages to upstream nixpkgs, I *might not be opposed to that, on a case by case basis.*

In general, I'm not convinced of the merit of these packages, all other technical considerations aside. So I'm not inclined to leave them.

File: pkgs/development/compilers/yosys/default.nix (hunk 13 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/default.nix]: The makefile was re-written at some point to make all of these changes obsolete; the Makefile now 1) attempts `clang` by default and 2) if that fails, falls back to `CONFIG := gcc`. Later on, iff `$(CONFIG) == gcc`, then `LD = gcc` (just grep `CONFIG` in the Makefile.)

So this all works out, as it should.

File: pkgs/development/compilers/yosys/default.nix (hunk 3 above)
Comment (CajuM) [reply pkgs/development/compilers/yosys/default.nix:47]:
```suggestion
  nativeBuildInputs = [ pkg-config protobuf flex bison python3 ];
  buildInputs = [ tcl readline libffi python3 protobuf zlib ];
```

protobuf and python3 are required at build-time

File: pkgs/top-level/python-packages.nix (hunk 12 above)
Comment (thoughtpolice) [reply pkgs/top-level/python-packages.nix:7408]: And yes, of all the packages, only `textx` seemed more generally useful at a glance. I could be convinced of others though, but that one I think could be added (that said I'm not sure how popular it actually *is*...)

**Comment by SuperSandro2000** [comment]
> the singular commit gives complete rational for the change and its design (vs scoping plugins in a different attrset or lambda) and splitting the two is pointless in light of this, even if it's a more standard form.

The name of the commit is at least wrong and generally we do one package per commit except for treewide changes. 

> Sorry, but no. I'm not about to get into debates about the merits of revert styles (of all things...), but that aside, I'll just ask: why would I want to revert 15 individual commits instead of just reverting 1?

Because you revert adding you as a maintainer and then adding you back. This will probably create some confusion and getting a merge conflict between those commits will probably be harder to resolve then individual reverts. You remove python packages we could keep and fine control is better than using a big bazooka. 

I would appreciate it if people would stop bikesheeding about almost everything lately.

File: pkgs/development/compilers/yosys/default.nix (hunk 13 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/yosys/default.nix]: This sounds awful. LD ?= gcc would be better.

**Comment by thoughtpolice** [comment]
> Because you revert adding you as a maintainer and then adding you back. This will probably create some confusion and getting a merge conflict between those commits will probably be harder to resolve then individual reverts.

It's completely fine. I re-added Mihai as a courtesy because I don't *not* want their contributions, and removing their name entirely felt nasty. For the most part, the other changes need to be re-thought, however, as I outlined elsewhere.

The point of crafting a set of changes is to *reduce confusion*. The intent of removing all these packages at once is to symbolize the fact I am undoing the entire merge, basically, *which is my intent*. If a conflict happens, that's life. I already reverted it; it's not a big deal.

Furthermore, re-adding some minor incidental changes from a series of patches that gets reverted is *extremely* standard protocol. I'm not trying to be an authority, but I'm also not exaggerating when I'm saying I've been doing some variation for this in FOSS projects for like, almost 15 years. It's completely run of the mill to largely revert changes, some minor conflicts notwithstanding, and incorporate small non-controversial ones. This is not weird or unusual, and it's very easy to understand. Again, I'm not sure what the point of this is other than to force your "help" onto this PR.

> You remove python packages we could keep and fine control is better than using a big bazooka.

Again: none of these packages meet the bar, IMO (as the maintainer of many related packages, and a user of many of them) to be included with the FPGA tooling in Nixpkgs, *yet, but maybe one day, they will*. I have repeatedly said this and outlined it in the original comment I linked, which I am assuming you have ignored completely.

To be clear, you seem to be suggesting I have "removed some packages that are useful and meet the bar for inclusion" or something, as if it was an oversight. On the contrary: I am explicitly arguing that none of them meet the bar for inclusion, and all of them should be removed. The only exception *might* be `textx`, which was only added as a dependency of all the others. I don't see a reason in keeping it when it supports nothing else.

> I would appreciate it if people would stop bikesheeding about almost everything lately.

Um, excuse me? *You* are the one literally on this PR, sitting here giving review comments requesting changes to my *whitespace formatting* saying things like, "I'm not a fan of this", as if you were going to be the one to maintain it (spoiler alert: you aren't.) I've given a rationale for everything I've done here, none of it is spurious or done without reason, and you have the gall to pull this one?

Frankly, despite the fact I have spent the better half of a decade working on Nixpkgs, and despite the huge utility of borg, stuff like this is exactly why I don't even bother making PRs these days, and today I did it as a courtesy. I'm honestly much better off routing around PRs from 99.9% of the users here by just committing directly to `master` like I usually do and ignoring your "advice" completely, which has already wasted a part of my morning.

If you don't like bikeshedding, allow me to make a suggestion: take a hint, review someone elses' packages.

**Comment by SuperSandro2000** [comment]
> I'm honestly much better off just committing to master, like I usually do.

If we take anything away from this than that you stop this behavior. It is not acceptable to push something to master except if it is an emergency fix.

**Comment by emilazy** [comment]
While I agree that it would be good to move to a model where everything is done through ofborg and PRs, to my knowledge there is no current consensus for adopting that as official policy and many maintainers push directly to master.

Please be more civil; @thoughtpolice is a long‐term contributor and maintainer and dropping argumentative reviews that only address trivial nitpicky style concerns isn’t helpful behaviour.

**Comment by emilazy** [comment]
If you care really strongly about highly uniform formatting in a codebase that already has a lot of micro‐level variation about such things then I’d suggest trying to improve [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt) to the point that it could be adopted as an automatically‐enforced standard.

**Comment by SuperSandro2000** [comment]
> many maintainers push directly to master.

I don't know from where you take that number but I just checked the last 4 pages of commits on master and I think almost all of them went through PRs if not all.

You don't need to agree with my formatting changes but I think we need to at least agree on that pushing to master is not a great thing to do. If you need your changes right now you should probably track a fork instead.

> I would appreciate it if people would stop bikesheeding about almost everything lately.

To maybe clarify this a bit: Lately I have the feeling that even about real changes people try to talk a bit more than usually.

> If you care really strongly about highly uniform formatting in a codebase that already has a lot of micro‐level variation about such things then I’d suggest trying to improve nixpkgs-fmt to the point that it could be adopted as an automatically‐enforced standard.

There is one thing I really would like to change in there which is that it shouldn't auto put everything in one line or one column but I have no clue about rust, ast or formatters and with my current time spend on reviewing and merging a lot I won't find the time to learn rust and fix that.

**Comment by thoughtpolice** [comment]
> If we take anything away from this than that you stop this behavior. It is not acceptable to push something to master except if it is an emergency fix. You have a pretty high chance to break evaluation or introduce something with a broken hash or a package that is buggy.

Sorry, if the choice is between "pushing to `master`" and "wasting a huge amount of time on a straightforward PR for packages I maintain over nothing", I'll be sure to take "push to `master`" every single time, with dessert, thanks.

> Also I suggested you some drive by whitespace changes which I am probably not even going to check if you just mark them as resolved.

Let me give a humble tip that, if you want to "entice" someone like me to file more PRs using the automation we have available, your first step can be to stop giving flippant "reviews" you don't even care to check up on yourself. Just a thought!

File: pkgs/development/compilers/yosys/default.nix (hunk 13 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/default.nix]: You can specify whatever variables you want with a custom `Makefile.conf`, but again, in this case the build is fine (I think the actual reason is because the `LDFLAGS` in yosys specifies `stdc++` anyway so the implicit inclusion it gets with the `g++/clang++` commands is matched, even if you only use `gcc/clang` to compile/link.)

File: pkgs/development/compilers/yosys/default.nix (hunk 5 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/default.nix:93]: Not committing this.

File: pkgs/development/compilers/yosys/default.nix (hunk 3 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/default.nix:47]: `python3` and `protobuf` must be linked to in this case (they are used at runtime), so leaving them in `buildInputs` is fine. `flex` and `bison` have been moved, however.

File: pkgs/development/compilers/yosys/default.nix (hunk 4 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/default.nix:51]: This should also have the `.patch` file moved out as well. Fixed.

File: pkgs/development/compilers/ghdl/default.nix (hunk 2 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/ghdl/default.nix:29]: Not accepting this one.

**Comment by SuperSandro2000** [comment]
> wasting a huge amount of time over nothing on a straightforward PR for packages I maintain

Maybe this is not a straightforward PR. If you mind taking a look at any straight forward update I probably merged a hundred or even more PRs without nitting to much in my opinion except maybe pkgconfig to pkg-config. I think most members of the community appreciate my efforts especially with testing on darwin and providing patches to fix simpler build failures.

> Let me give a humble tip that, if you want to entice someone like me to file more PRs using the automation we have available, your first step can be to stop giving flippant "reviews" you don't even care to check up on yourself because the matter is so utterly inconsequential. Just a thought!

So how do you want to improve the overhaul quality of packages if in your opinion I am not allowed to give formatting feedbacks that slightly improve readability and are nice to have. Should I just apply the changes and directly push them to master? Or write an opinionated formatter that I enforce on everyone? I don't think that those thoughts would be any better.

I've contributed a not to ignore part to nixpkgs in form of commits and getting down to 2.2 open PRs slowly but steadily. I would appreciate it if you would be a bit more friendly. We all want a pleasant community for everyone and I try to calm this down but right now you are just being a bit of a jerk instead of finding an agreement that works for you and for me.

**Comment by thoughtpolice** [comment]
> Maybe this is not a straightforward PR. If you mind taking a look at any straight forward update I probably merged a hundred or even more PRs without nitting to much in my opinion except maybe pkgconfig to pkg-config. I think most members of the community appreciate my efforts especially with testing on darwin and providing patches to fix simpler build failures.

If you feel belittled I truly apologize, and that's my fault for being frustrated and grumpy this morning. But truly and honestly, I don't see what isn't straightforward about all of this! Again, I believe I have explained the motivation behind these changes fully. These packages were, I believe, added in error, and are better located elsewhere. On top of that, I made a minor change to the way yosys loads plugins to match the way it behaves for non-Nix users, today (which is coincidentally a feature the original PR fixed, in a different way, that I did not think was correct.) I included an example of this new feature. There's really nothing confusing about any of it, if you ask me.

(In fact, the reason I didn't just commit this series to `master` isn't because it's complicated or anything; it's because I just wanted to test evaluation and also give a pointer to 2660037f8cdf6bb69bacbf3d9d2f9f72d056c8d6 to Mihai, for reference.)

(And if it's any consolation, I respect anyone who tests and helps fix things on Darwin, and it means you clearly do have more patience than I do.)

> So how do you want to improve the overhaul quality of packages if in your opinion I am not allowed to give formatting feedbacks that slightly improve readability and are nice to have. Should I just apply the changes and directly push them to master? Or write an opinionated formatter that I enforce on everyone? I don't think that those thoughts would be any better.

Well, I think the first thing is that you have to realize that general quality of Nixpkgs is not held down by things like syntactical formatting of Nix code. That's all pretty minor. It can be ignored, and in many cases, it largely only serves to demoralize would-be contributors. It just feels nitpicky. I don't think it's a good way to conduct ourselves here, and I think doing it leads to a kind of weird alienation that we're seeing right now: I've literally been contributing to this project for almost 10 years. I was hoping to have this whole issue done with many hours ago, but now I'm being asked to change whitespace? For what? None of this is substantial to any user of the package, it's just "busy work" we're creating for ourselves, it feels like. But I know you're just being honest and speaking truly that you think this is an improvement, and yes, we all have to be good citizens (hell, whenever I take over packages that get abandoned, I often reformat them to my liking, after the fact. I get it.)

Which leads to my next point that I'll try to explain, and which is probably the core of the whole thing going on here: I am almost never, unless explicitly noting otherwise, opening PRs to actually get them "reviewed" by anyone. Almost never! If anyone does review it, it's often only for small changes that I apply as a courtesy. I am only using the review process to trigger borg, get a clear on my eval/build status, merge, and move on. This sounds insane at first glance, especially to you, who clearly spends a lot of time reviewing PRs, but I assure you, it isn't malicious. So, why would I do this? Because:

  - I am almost always PRing packages that I already maintain. Realistically, I am the most qualified person to even review the change in the first place. There is often very little feedback that can be given, beyond minor things. Most of these are small and do not actively improve the quality of the package for users, or are purely aesthetic in nature.
  - As evidence of this, the vast majority of changes you've recommended here are mostly pretty small an inconsequential. I note only the change to use `lib` instead of `stdenv.lib` as part of a recent tree-wide change. This isn't unusual. Most of the changes people request on PRs like this &mdash; me modifying my own packages &mdash; are pretty small. It's not an accusation or fault; it's just the reality that I've fixed most of the kinks, and I'm the only party significantly invested in them. So there's not much input for other people to give! And it's not like anyone is going to keel over if `flex` is in `buildInputs` instead of `nativeBuildInputs`.
  - Finally: I want to cycle and test evaluation with borg, and especially test with Darwin if applicable. (While I don't go out of my way to *support* Darwin, I try to ensure it does not break, if it worked before, and borg is the easiest way to do that right now.)

So this whole thing comes down to that. I don't actually *want* review input, generally, because most of the time, I'm the only maintainer anyway. Maybe I should make this more clear somehow; clearly it's a communication failure on my part. I don't say this to antagonize anyone, and it doesn't mean I want *no* input ever, but I want actionable input that isn't just superfluous. As I hope you noticed, I did incorporate several of your changes that you requested including the `lib` changes and all the `buildInput/postPatch` fixes (and credited you; I hate using @'s in commits!) These are all just small correctness/accepted changes, and they're all fine.

Now, circling back to your question: Should you just merge this? Should you use an autoformatter? Neither. My honest answer, were I in your shoes, would just be "Leave it alone, maybe give a small approval, move on." Seriously! I actually try not to merge things on behalf of other committers; I trust they'll merge it when they're ready, and I think just giving a small approval is courteous. I don't often understand the packages if I give a courtesy review, or only at a high level, because I'm often not a user. So I can't often give substantial feedback, and don't. I might give small suggestions like you did about buildInputs, but, giving lots of "extraneous" reviews consisting of formatting, syntax issues just feels nit-picky. I also don't think autoformatters are the decisions of individuals: Nix contributors must collectively decide whether to apply rules like this to the whole tree. If such a tool were accepted, even were I not to vote on it, I'd acknowledge it. It might be a good idea to try that in the future; if a robot told me "Format this whitespace" I might not take it as I did earlier, even if they were both rules.

Doing adequate code reviews is difficult because there are often many hidden intrapersonal assumptions behind a simple PR in a tab on Github, like here. I clearly should have set the expectations better, I guess. But I think a lesson you might want to take from this is not to make people sweat the tiny stuff. It not only is alienating (even, and maybe especially, to experienced contributors like myself) but I learned a long time ago that it largely serves little purpose: while it might appear to instill rigor, it only instills rigidity. And it's such a waste of time! I think having minor imperfections like this isn't anything to sweat over. We can always fix it later, and even if we don't, there's nothing to cry over.

So how *would* I suggest you improve the quality on your own? I dunno, honestly. I don't think any one person can easily do that, and even if you could, I don't think formatting changes and stuff like that would be the way. I think the quality of Nix is constantly improving, but only due to painstaking work from many, many people (including you.) But you're obviously doing a much better job than I am, so maybe you should just ignore me entirely.

> I've contributed a not to ignore part to nixpkgs in form of commits and getting down to 2.2 open PRs slowly but steadily. I would appreciate it if you would be a bit more friendly. We all want a pleasant community for everyone and I try to calm this down but right now you are just being a bit of a jerk instead of finding an agreement that works for you and for me.

I do appreciate that, and thank you. Nixpkgs is only useful to me because everybody else does "all the other stuff". Frankly I've had a rough winter and have not been in a particularly good mood recently, so I squarely admit to being a jerk and apologize for it (though I understand if you wouldn't accept.) Having said all that, I believe I've adequately addressed all your review criteria as well, aside from the ones I simply disagree with relating to formatting and whitespace. I'll refrain from being an ass in the future, though I think most if your requests here are dealt with, unless there's anything else.

**Comment by SuperSandro2000** [comment]
Apology accepted. 

> And it's not like anyone is going to keel over if flex is in buildInputs instead of nativeBuildInputs.

If I remember correctly this is very important if you are cross compiling. If the inputs are not correct you might execute the binary in the architecture for the target instead of the host.

I think most of the formatting changes are coming from the revert which threw me very much of that you undid things intentionally.

**Review by SuperSandro2000 — APPROVED** [review]
Diff LGTM

**Comment by jonringer** [comment]
> And it's not like anyone is going to keel over if flex is in buildInputs instead of nativeBuildInputs.

It's been a long term goal for us to switch `strictDeps = true;`, which would mean that buildInputs would stop adding items to `PATH`. In this view, @SuperSandro2000's suggestions can be considered "more correct, and aligned with best practices".

Although some of the suggestions can seem as nit-picky, nixpkgs is huge, and having more consistency across packages does enable the ability to parse the packages more easily.

**Comment by jonringer** [comment]
> Doing adequate code reviews is difficult because there are often many hidden intrapersonal assumptions behind a simple PR in a tab on Github, like here.

In general I find this be untrue, most package ecosystems follow pretty set standard around their toolchains. Python, go, rust, autotools, meson, and cmake builds are usually similar to other builds using the same toolchain. So, you can still have people review and improve your PR in some aspects, without having to know the entirety of a package's context.

**Comment by thoughtpolice** [comment]
@ofborg build yosys-bluespec yosys-ghdl

**Comment by thoughtpolice** [comment]
Sigh, Darwin and aarch64 are backed up as usual, but x86_64-linux went through...

File: pkgs/development/compilers/yosys/default.nix (hunk 13 above)
Comment (Luflosi) [reply pkgs/development/compilers/yosys/default.nix]: This change broke compilation on darwin (see https://github.com/NixOS/nixpkgs/pull/112950/commits/5bf5284c21af27cdd870137cf76428f3e0b78a1b) or maybe it was already broken by an earlier change, I did not bisect that far. But if I re-add `--replace 'CXX = clang' "" \`, it compiles again. Why was this removed? Can I add it back? Should I only add this one line since that seems to be all that is needed? Should I make this a conditional that only adds this line on dariwn? I tested adding the line on NixOS and there were no changes in the output except for a different hash.
